services/sso/src/main/java/software.amazon.awssdk.services.sso/internal/SsoCredentialsUtils.java

```diff
@@ -0,0 +1,165 @@
+                                                              + " otherwise invalid. To refresh this SSO session run aws sso"
+                                                              + " login with the corresponding profile.").build();
+            }
+
+        } catch (IOException e) {
+            throw SdkClientException.builder().message("Unable to read the cached token file. Please check if the correct "
+                                                       + "sso_start_url is configured in the profile").cause(e).build();
+        }
+
+        return getCredentialsFromSso();
+    }
+
+    private void validateAndResolveSsoProfile(Profile profile) {
+        Map<String, String> configurations = profile.properties();
+
+        if (configurations.containsKey(SSO_ACCOUNT_ID)) {
+            ssoAccountId = configurations.get(SSO_ACCOUNT_ID);
+        } else {
+            throw SdkClientException.create(String.format("This profile is invalid, %s is missing.", SSO_ACCOUNT_ID));
+        }
+
+        if (configurations.containsKey(SSO_REGION)) {
+            ssoRegion = configurations.get(SSO_REGION);
+        } else {
+            throw SdkClientException.create(String.format("This profile is invalid, %s is missing.", SSO_REGION));
+        }
+
+        if (configurations.containsKey(SSO_ROLE_NAME)) {
+            ssoRoleName = configurations.get(SSO_ROLE_NAME);
+        } else {
+            throw SdkClientException.create(String.format("This profile is invalid, %s is missing.", SSO_ROLE_NAME));
+        }
+
+        if (configurations.containsKey(SSO_START_URL)) {
+            ssoStartUrl = configurations.get(SSO_START_URL);
+        } else {
+            throw SdkClientException.create(String.format("This profile is invalid, %s is missing.", SSO_START_URL));
+        }
+
+    }
+
+    /**
+     * Call SSO to retrieve a new {@link AwsSsoCredentials}.
+     * @return
+     */
+    private AwsSsoCredentials getCredentialsFromSso() {
+        GetRoleCredentialsRequest request = GetRoleCredentialsRequest.builder().accountId(ssoAccountId)
+                                                                               .roleName(ssoRoleName)
+                                                                               .accessToken(tokenProperties.getAccessToken())
+                                                                               .build();
+        GetRoleCredentialsResponse response = ssoClient.getRoleCredentials(request);
+        RoleCredentials roleCredentials = response.roleCredentials();
+        return AwsSsoCredentials.create(roleCredentials.accessKeyId(), roleCredentials.secretAccessKey(),
+                                        roleCredentials.sessionToken(), roleCredentials.expiration());
+
+    }
+
+    private boolean tokenExpired(String expiresTime) {
+        return Instant.now().isAfter(Instant.parse(expiresTime));
```

Comment:
Do we not want to treat expiring soon as expired?

```
When checking to see if a cached token is expired, the SDK MAY treat the token as expired if it will expire within 15 minutes of the expiresAt time
```